### PR TITLE
cargo/config: disallow build warnings

### PR DESCRIPTION
**Pull request by dianpopa**
## Reason for This PR

Treat build errors as warnings.

## Description of Changes

Changed cargo/config.
**Obs** Tests are expected to fail until #1693 gets merged.

- [x] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.

### CODE REVIEW

File: .cargo/config [hunk 1]
```diff
@@ -8,5 +8,11 @@ target-dir = "build/cargo_target"
 # No need for the `crt_static` flag as rustc appends it by default.
 rustflags = [
 	"-C", "link-arg=-lgcc",
+	"-D", "warnings",
 	"-C", "link-arg=-lfdt",
 ]
+
+[target.'cfg(not(any(target_arch="arm", target_arch="aarch64")))']
```
Comment (andreeaflorescu) [inline .cargo/config:15]: This can easily become annoying when doing development. If you have code in flux and you want to test it, my understanding is that the build will fail locally. Can we instead have this as a separate test in the CI to check that when trying to push something upstream there are no build warnings?

Comment (acatangiu) [reply .cargo/config:15]: Can we do this in cargo for release builds only?

Comment (dianpopa) [reply .cargo/config]: I moved the check inside the integration tests. Take a look!

File: tests/host_tools/cargo_build.py [hunk 2]
```diff
@@ -70,3 +70,11 @@ def get_firecracker_binaries():
     jailer_bin_path = "{}/{}".format(out_dir, JAILER_BINARY_NAME)
 
     return fc_bin_path, jailer_bin_path
+
+
+def get_rustflags():
+    """Get the relevant rustflags for building/unit testing."""
+    rustflags = "-D warnings"
+    if platform.machine() == "aarch64":
+        rustflags += " -C link-arg=-lgcc -C link-arg=-lfdt "
```
Comment (iulianbarbu) [inline tests/host_tools/cargo_build.py:79]: If I understand correctly, we would link against `-lgcc` and `-lfdt` for the firecracker binary used inside the integration tests and when cargo testing on arm. Why is this needed now, when we disallow warnings?

Comment (dianpopa) [reply tests/host_tools/cargo_build.py:79]: We need this flag for aarch64 binary to link (these flags are always used when doing `cargo build` on aarch64. Take a look at the cargo/.config for arm64. Unfortunately there is no way (or I do not know any) of getting the flags set in cargo/.config for some platform from our python tests.

Comment (iulianbarbu) [reply tests/host_tools/cargo_build.py:79]: Ok, so when compiling with `cargo build`, using `TARGET_CC=musl-gcc`, this has to link on aarch64 against `-lgcc` and `-lfdt` always (because musl-gcc lacks some symbols used in firecracker), and these flags can be found in `$HOME/.cargo/config`. I've looked on my x86 machine, and found this config file inside the .cargo dir inside the project, as well as this part responsible for aarch64 specific linking flags.

It makes sense, once we compile using rustc, and not `cargo build`, to specify these flags in the command line.

Comment (iulianbarbu) [reply tests/host_tools/cargo_build.py:79]: Oh, actually, looked once more, and found out that we actually use `cargo build` to build the firecracker binaries. Using a .cargo/config with this directive, inside the project where this build happens, should be enough:

[target.'cfg(any(target_arch="arm", target_arch="aarch64"))']
rustflags = [
        "-C", "link-arg=-lgcc",
        "-C", "link-arg=-lfdt",
]

However, this is more like a nit now. I do not have a strong preference, to use this or what you've already pushed in this PR.

**Comment by dianpopa** [comment]
@iulianbarbu @serban300 please re-review. Tests have passed!
